source/agora/test/GossipProtocol.d

```diff
@@ -0,0 +1,237 @@
+/*******************************************************************************
+
+    Contains tests for Gossip Protocol.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.GossipProtocol;
+
+import agora.common.Data;
+
+import agora.protocol.GossipProtocol;
+
+import std.digest.sha;  
+ 
+ /// Test API
+interface TestAPI {
+
+    public bool setMessage(Hash msg);
+}
+
+/// Remote Node for test
+class TestRemoteNode
+{
+    /// API client to the node
+    private TestAPI api;
+
+    /// Ctor
+    public this (TestAPI api)
+    {
+        this.api = api;
+    }
+
+    /// Send a message
+    public bool sendMessage(Hash msg) {
+        return this.api.setMessage(msg);
+    }
+}
+
+/// Network for test
+class TestNetwork {
+    private TestRemoteNode[] remoteNodes;
+
+    /// Ctor
+    public this () {
+    }
+
+    public void setRemotes (TestAPI [] api) {
+        for (int idx = 0; idx < api.length; idx++) {
+            this.remoteNodes ~= new TestRemoteNode(api[idx]);
+        }
+    }
+
+    /// Send a message only to the node passed by the filter delegate.
+    public void sendMessage(Hash msg) 
+    {
+        foreach(node; this.remoteNodes) {
+            node.sendMessage(msg);
+        }
+    }
+}
+
+/// Local Node for test
+class TestNode : TestAPI, IProtocolOwner {
+    private string name;
+    private TestNetwork network;
+    private GossipProtocol gossip;
+
+    public Hash lastReceivedMessage;
+    public int receivedMessageCount;
+
+    public this (string name) {
+        this.name = name;
+        this.network = new TestNetwork();
+        this.gossip = new GossipProtocol(this);
+
+        this.receivedMessageCount = 0;
+    }
+
+    /// PUT: /message
+    public bool setMessage(Hash msg) {
+		bool res = this.gossip.receiveMessage(msg);
+        if (res) {
+		    this.receivedMessageCount++;
+            this.lastReceivedMessage = msg;            
+        }
+        return res;
+    }
+
+    /// Set Romot Node
+    public void setRemotes (TestAPI [] api) {
+        this.network.setRemotes(api);
+    }
+
+    /// Send a message only to the node passed by the filter delegate.
+    public void sendMessage(Hash msg) {
+        this.network.sendMessage(msg);
+    }
+}
+
+unittest {
```

Comment:
This is a good start. There's a few more tests I can think of to add:

- test sending multiple messages, and not only one.
- test sending the same message multiple times and assert on the return value of `setMessage`. The first time `sha256Of("Message No. 1")` is sent to a node, it should return `true` because it's a new item. On the second time `sha256Of("Message No. 1")` is sent to the same node, it should return `false`.

Reply:
Thank you for your good opinion.
I'll add it and update it.

Reply:
Modified.

Reply:
Could you add a test-case I mentioned above in the second point? Send the same message twice and assert that `setMessage` returns `true` the first time, and `false` the second time.

Reply:
That's a great idea!

Reply:
Added

Reply:
@AndrejMitrovic 
Check Please~